### PR TITLE
feat(yarn): improve workspace support

### LINKED ISSUE #880: `yarn workspace` does not support complete package `workspaces` specification syntax
### Sanity checks

- [X] My issue relates to a specific CLI completion spec (e.g. `git checkout` is missing options in `git` completion spec). If your issue is more general, please create your issue here: [withfig/fig](https://github.com/withfig/fig/issues/new?assignees=&labels=NEED_TO_LABEL&template=main_issue_template.yml)
- [X] I have searched [github.com/withfig/autocomplete/issues](https://github.com/withfig/autocomplete/issues?q=) and there are no duplicates of my issue
- [ ] I would like to work on this.

### What CLI tool does this relate to?

yarn

### Which statement makes the most sense?

This completion spec isn't complete

### Issue Details

It looks like support for `yarn workspace` was most recently enhanced in [this PR](https://github.com/withfig/autocomplete/pull/678) however it does not yet support the complete specification. It correctly generates workspace names if the root `package.json` uses this format:

```json
{
  "workspaces": [
    "packages/*",
    "plugins/*",
    "sites/*"
  ]
}
```
...but it does not pick up workspace names if the root `package.json` is using the more complex format here:
```json
{
  "workspaces": {
    "packages": [
      "packages/*",
      "plugins/*",
      "sites/*"
    ],
    "nohoist": ["**/react-native", "**/react-native/**"]
  }
}
```

### PR DESCRIPTION
**Pull request by elevatebart**
### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Bug fix

### What is the current behavior? (You can also link to an open issue here)

Closes #880

### What is the new behavior (if this is a feature change)?

`yarn workspace` now works with the full specification
A good testing repo to test this is [cypress](https://github.com/cypress-io/cypress)

### Details

- I used a slightly **more functional** version of the idea of the previous pull request.
- I improved reactivity by using `generateSpec`. It only resolves the scripts when the package is called.

### CODE REVIEW

**Comment by elevatebart** [comment]
Follow-up to @bolshoytoster PR 

- #887

**Review by fedeci — CHANGES_REQUESTED** [review]
This doesn't work for yarn@2 or yarn@3 ([reference](https://github.com/withfig/autocomplete/pull/887#pullrequestreview-866993143))

File: src/yarn.ts [hunk 1]
```diff
@@ -1375,78 +1352,60 @@ const completionSpec: Fig.Spec = {
...
-                subcommands.push({
-                  name: await getPackageName(workspace),
-                  description: "Workspaces",
+          const subcommands = Object.entries(workspacesDefinitions).map(
+            ([name, workspaceDef]: [string, { location: string }]) => ({
+              name,
+              description: "Workspaces",
+              generateSpec: async () => {
+                return {
+                  name: `${name} - scripts`,
```
Comment (fedeci) [inline src/yarn.ts:1371]: This `generateSpec` is not required and the `postProcess` can be added directly to the outer `generateSpec` args.

File: src/yarn.ts [hunk 2]
```diff
@@ -1375,78 +1352,60 @@ const completionSpec: Fig.Spec = {
...
                   args: {
                     name: "script",
                     generators: {
-                      script: `\cat ${workspace}/package.json`,
-                      postProcess,
+                      cache: {
+                        strategy: "stale-while-revalidate" as const,
```
Comment (fedeci) [inline src/yarn.ts:1376]: Why is `as const` required here?

**Review by fedeci — CHANGES_REQUESTED** [review]
Thanks for this PR @elevatebart!

File: src/yarn.ts (hunk 2 above)
Comment (elevatebart) [reply src/yarn.ts:1376]: TypeScript refuses to consider `"stale-while-revalidate"` as a constant unless you tell it. 
By default, it considers it a modifiable string.

File: src/yarn.ts (hunk 1 above)
Comment (elevatebart) [reply src/yarn.ts]: Fixed it ;)

File: src/yarn.ts (hunk 2 above)
Comment (fedeci) [reply src/yarn.ts]: That's super strange. It is statically typed and ts should recognize it correctly as an accepted string.

Comment (elevatebart) [reply src/yarn.ts]: It started mattering in typescript 4 for enum literal types.
I could also type the whole created object if you prefer.

Comment (elevatebart) [reply src/yarn.ts]: I did this in 6da6b53

**Comment by fedeci** [comment]
Still not working for v2 and v3, see the examples below:
- For yarn@1 this is the output of `yarn workspaces info --silent`
```json
{
  "@fig/autocomplete-generators": {
    "location": "packages/autocomplete-generators",
    "workspaceDependencies": [
      "@withfig/autocomplete-types"
    ],
    "mismatchedWorkspaceDependencies": []
  },
  "@fig/autocomplete-helpers": {
    "location": "packages/autocomplete-helpers",
    "workspaceDependencies": [
      "@withfig/autocomplete-types"
    ],
    "mismatchedWorkspaceDependencies": []
  },
  "@fig/autocomplete-merge": {
    "location": "packages/autocomplete-merge",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  },
  "@withfig/autocomplete-tools": {
    "location": "packages/autocomplete-tools",
    "workspaceDependencies": [
      "@withfig/autocomplete-types"
    ],
    "mismatchedWorkspaceDependencies": [
      "@withfig/commander",
      "@fig/autocomplete-helpers",
      "@fig/autocomplete-merge"
    ]
  },
  "@withfig/autocomplete-types": {
    "location": "packages/autocomplete-types",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  },
  "@fig/eslint-config-autocomplete": {
    "location": "packages/eslint-config-autocomplete",
    "workspaceDependencies": [
      "@withfig/eslint-plugin-fig-linter"
    ],
    "mismatchedWorkspaceDependencies": []
  },
  "@withfig/eslint-plugin-fig-linter": {
    "location": "packages/eslint-plugin-fig-linter",
    "workspaceDependencies": [],
    "mismatchedWorkspaceDependencies": []
  },
  "@withfig/oclif": {
    "location": "packages/oclif",
    "workspaceDependencies": [
      "@withfig/autocomplete-types"
    ],
    "mismatchedWorkspaceDependencies": []
  },
  "@withfig/commander": {
    "location": "packages/commander",
    "workspaceDependencies": [
      "@withfig/autocomplete-types"
    ],
    "mismatchedWorkspaceDependencies": []
  }
}
```

- For yarn@2 and yarn@3 this is the output of `yarn workspaces list --json`
```sh
{"location":".","name":"babel"}
{"location":"benchmark","name":"@babel/benchmark"}
{"location":"codemods/babel-plugin-codemod-object-assign-to-object-spread","name":"@babel/plugin-codemod-object-assign-to-object-spread"}
{"location":"codemods/babel-plugin-codemod-optional-catch-binding","name":"@babel/plugin-codemod-optional-catch-binding"}
{"location":"eslint/babel-eslint-config-internal","name":"@babel/eslint-config-internal"}
{"location":"eslint/babel-eslint-parser","name":"@babel/eslint-parser"}
{"location":"eslint/babel-eslint-plugin","name":"@babel/eslint-plugin"}
{"location":"eslint/babel-eslint-plugin-development","name":"@babel/eslint-plugin-development"}
{"location":"eslint/babel-eslint-plugin-development-internal","name":"@babel/eslint-plugin-development-internal"}
{"location":"eslint/babel-eslint-shared-fixtures","name":"@babel/eslint-shared-fixtures"}
{"location":"eslint/babel-eslint-tests","name":"@babel/eslint-tests"}
{"location":"packages/babel-cli","name":"@babel/cli"}
{"location":"packages/babel-code-frame","name":"@babel/code-frame"}
{"location":"packages/babel-compat-data","name":"@babel/compat-data"}
{"location":"packages/babel-core","name":"@babel/core"}
{"location":"packages/babel-generator","name":"@babel/generator"}
{"location":"packages/babel-helper-annotate-as-pure","name":"@babel/helper-annotate-as-pure"}
```

As you can see they changed both the format, the command and the flags, I would get the correct output relative to the used version, convert it to a common format like `type Workspace = { name: string, path: string }` and then generate subcommands and args relative to that.

File: src/yarn.ts [hunk 3]
```diff
@@ -1375,78 +1353,72 @@ const completionSpec: Fig.Spec = {
       name: "workspace",
       description: "Manage workspace",
       generateSpec: async (_tokens, executeShellCommand) => {
-        const { postProcess } = scriptList;
-        const subcommands = [];
+        const version = await executeShellCommand("yarn --version");
+        const isYarnV1 = /^1\.\d+\.\d+/.test(version);
```
Comment (fedeci) [inline src/yarn.ts:1357]: nit: this is more performant
```suggestion
        const isYarnV1 = version.startsWith('1');
```

File: src/yarn.ts [hunk 4]
```diff
@@ -1375,78 +1353,72 @@ const completionSpec: Fig.Spec = {
...
 
-          if (out.trim() == "") {
-            return { name: "workspaces" };
-          }
-          const packageContent = JSON.parse(out);
-          const workspaces = packageContent["workspaces"];
+          const workspacesDefinitions = isYarnV1
+            ? // transform Yarn V1 output to array of workspaces like Yarn V2
+              Object.entries(
+                JSON.parse(out) as Record<string, { location: string }>
```
Comment (fedeci) [inline src/yarn.ts:1370]:
```suggestion
                JSON.parse(out.slice(out.indexOf("\n") + 1, out.lastIndexOf('\n'))) as Record<string, { location: string }>
```

File: src/yarn.ts [hunk 5]
```diff
@@ -1375,78 +1353,72 @@ const completionSpec: Fig.Spec = {
...
-            );
+          const subcommands: Fig.Subcommand[] = workspacesDefinitions.map(
+            ({ name, location }: { name: string; location: string }) => ({
+              name,
+              description: "Workspaces",
+              args: {
+                name: "script",
+                generators: {
+                  cache: {
+                    strategy: "stale-while-revalidate",
```
Comment (fedeci) [inline src/yarn.ts:1391]:
```suggestion
                    strategy: "stale-while-revalidate",
                    ttl: 60_000, // 60s
```

**Review by fedeci — CHANGES_REQUESTED** [review]
Here are some other suggested changes. We definitely need to implement cache for `generateSpec` too, for the moment this is great!

File: src/yarn.ts (hunk 4 above)
Comment (elevatebart) [reply src/yarn.ts:1370]: What would be a good test for this change?

**Comment by elevatebart** [comment]
@fedeci one last question I think we are ready to go.

File: src/yarn.ts (hunk 4 above)
Comment (fedeci) [reply src/yarn.ts]: Any yarn@1 workspace. There are some header and footer information printed along with the json.

**Review by fedeci — APPROVED** [review]
Thanks! Great work @elevatebart!

File: src/yarn.ts (hunk 4 above)
Comment (elevatebart) [reply src/yarn.ts]: not with the `--silent` option (at least from what I saw)

Comment (elevatebart) [reply src/yarn.ts]: OK it does...

Comment (fedeci) [reply src/yarn.ts:1369]: 😅yeah

**Pull request merged**